### PR TITLE
[WIP] automated linearity checks

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Currently blocked by https://github.com/JuliaSymbolics/SymbolicUtils.jl/issues/94

### CODE REVIEW

**Comment by shashi** [comment]
@ChrisRackauckas could you add tests for the other 2 systems?

Right now the `@btime` for https://github.com/SciML/ModelingToolkit.jl/pull/382/files#diff-976b86e531528452332bc85d74807c8dR17 is ~100ms, it could be faster I guess. but we can look at it if it becomes a bottleneck.

**Comment by shashi** [comment]
This can be done using `hessian_sparsity` now....

**Comment by ChrisRackauckas** [comment]
Oh yes, hadn't thought of that! Could you whip up a version and we can just quickly merge then?

File: src/direct.jl [hunk 1]
```diff
@@ -100,41 +100,57 @@ end
...
+                  b = isidx(~y) ? ~y : _scalar
                   combine_terms(linearity(~f, Val{2}()), a, b)
               else
                   error("Function of unknown linearity used: ", ~f)
               end
-          end]
+          end] |> Rewriters.Chain |> Rewriters.Postwalk |> Rewriters.Fixpoint
+
+    # we do this in a let block so that Revise works on the list of rules
+    global hessian_sparsity
```
Comment (ChrisRackauckas) [inline src/direct.jl:127]: what's this for?

Comment (ChrisRackauckas) [reply src/direct.jl:127]: oh because of the let block?

Comment (shashi) [reply src/direct.jl:127]: yeah.

**Pull request merged**